libsolidity/ast/Types.cpp

```diff
@@ -345,7 +345,7 @@ TypePointer IntegerType::binaryOperatorResult(Token::Value _operator, TypePointe
 		return TypePointer();
 
 	// All integer types can be compared
-	if (Token::isCompareOp(_operator))
+	if (Token::isCompareOp(_operator) || Token::isShiftOp(_operator))
```

Comment:
We can choose to use the type of the left operand.  I don't see a reason why shifting a number should change the type of the number.  For me it's more natural if `uint8 << int32` stays to be `uint8` (why does a sign flag suddenly appear?).